src/Compilers/CSharp/Test/Emit/Emit/DynamicAnalysis/DynamicInstrumentationTests.cs

```diff
@@ -2120,5 +2118,159 @@ private CompilationVerifier CompileAndVerify(string source, string expectedOutpu
+
+        class InstrumentationChecker
+        {
+            StringBuilder consoleExpectations = new StringBuilder();
+            Dictionary<int /* method */, MethodChecker> spanExpectations = new Dictionary<int, MethodChecker>();
+
+            public InstrumentationChecker()
+            {
+                consoleExpectations.AppendLine($"Flushing");
+            }
+
+            /// <summary>
+            /// Start recording expectations for a method.
+            /// </summary>
+            public MethodChecker M(int method, int file, string snippet)
+            {
+                consoleExpectations.AppendLine($"Method {method}");
+                consoleExpectations.AppendLine($"File {file}");
+
+                var result = new MethodChecker(consoleExpectations).T(snippet);
+                spanExpectations.Add(method, result);
+                return result;
+            }
+
+            /// <summary>
+            /// Start recording expectations for a method, but without checking snippets for this method.
+            /// </summary>
+            public MethodChecker M(int method, int file)
+            {
+                consoleExpectations.AppendLine($"Method {method}");
+                consoleExpectations.AppendLine($"File {file}");
+
+                return new MethodChecker(consoleExpectations, noSnippets: true).T();
+            }
+
+            public string ExpectedOutput { get { return consoleExpectations.ToString(); } }
+
+            /// <summary>
+            /// Verify the recorded expectations.
+            /// </summary>
+            internal void CompleteCheck(Compilation compilation, string source)
+            {
+                var peImage = compilation.EmitToArray(EmitOptions.Default.WithInstrumentationKinds(ImmutableArray.Create(InstrumentationKind.TestCoverage)));
+                var peReader = new PEReader(peImage);
+                var reader = DynamicAnalysisDataReader.TryCreateFromPE(peReader, "<DynamicAnalysisData>");
+                string[] sourceLines = source.Split('\n');
+
+                if (spanExpectations.Count == 0)
+                {
+                    Dump(reader, sourceLines);
+                    return;
+                }
+
+                foreach (int method in spanExpectations.Keys)
+                {
+                    var actualSnippets = GetActualSnippets(method, reader, sourceLines);
+                    var expectedSnippets = spanExpectations[method].SnippetExpectations;
+
+                    if (actualSnippets.Length != expectedSnippets.Length)
+                    {
+                        Assert.False(true, $"The number of spans was expected to be {expectedSnippets.Length} for the method {method}, but it actually has {actualSnippets.Length}.");
+                    }
+
+                    for (int index = 0; index < actualSnippets.Length; index++)
```

Comment:
The scenarios: the initial test source and when something breaks the test show before and after and a diff.

```AssertEx.Equal``` is pretty powerful -- it implements basic line-by-line differ. 
http://source.roslyn.io/#System.Reflection.Metadata.Tests/TestUtilities/AssertEx.cs,95a4e29ac2ebf80c

It's used for IL and PDB tests. The idea is basically that when comparing sequences of things it does the comparison and if is not equal it prints out results (with a diff) using a custom printer. So you can craft it to print out the actual and expected source code of the test that can be copy-pasted to the actual source.